src/storage/engine/mod.rs

```diff
@@ -101,6 +103,24 @@ pub trait Engine: Send + Debug {
         }
     }
 
+    fn future_snapshot(
+        &self,
+        ctx: &Context,
+    ) -> Box<Future<Item = Box<Snapshot>, Error = Error> + Send> {
```

Comment:
How about using `conservative_impl_trait`?

Reply:
Cannot use `conservative_impl_trait` here. First of all, it is in a trait, and as far as I known, `impl Trait` in trait is not implemented.

Secondly, I failed to use `impl Trait` syntax here even if it is not in a trait. I guess it is because`future::err` and `future. ... .flatten()` have different return types.

Reply:
So it looks like adding an extra heap allocation.

Reply:
Yes. The only way to avoid it is to implement some `struct SnapshotFuture` just like other future structs, which is a bit complex and does not have significant improvements.